### ISSUE TITLE
Implicit UTF-8 conversion for binary payload of MQTT messages

### ISSUE THREAD

**Issue by DurandA** [node-red/node-red#1912]
When receiving message from the broker, the *mqtt input* tries to parse it to a string:
https://github.com/node-red/node-red/blob/e03a0fffa9ea2b195d5acda5a8e918b148bce1b7/nodes/core/io/10-mqtt.js#L401

This is very unfortunate when working with binary messages. It could happen that the payload is a valid UTF-8 string.

Would you accept a PR adding a checkbox that can be ticked to ensure that the payload is kept as a Buffer object?

**Comment by knolleary** [comment]
Hi, the issue template you've deleted describes our contribution process for feature suggestions and similar requests.

```
If your issue is:
  - a general 'how-to' type question,
  - a feature request or suggestion for a change,
  - or problems with 3rd party (`node-red-contrib-`) nodes

please use the [Node-RED Forum](https://discourse.nodered.org) or [slack team](https://nodered.org/slack).
```

As you can see, the code attempts to do the right thing, and does work in the vast majority of cases.

I think a checkbox to require a specific encoding is a reasonable idea. The main thing is to ensure 100% backwards compatibility.

**Comment by DurandA** [comment]
Sorry, the issue template did not appear as I opened the issue from a source code file.

I will make a commit and then discuss about the change on Slack.

**Comment by knolleary** [comment]
Given we already have this issue, I don't mind using it for this discussion. You've passed the first level of filtering out things we wouldn't consider accepting 😁

**Comment by DurandA** [comment]
I made a preliminary version on DurandA/node-red@84a6a52. This is mostly copy-pasted from the udp input node.

Please let me know if you have some comments and if you would accept it as a PR.

**Comment by knolleary** [comment]
@DurandA I do have a couple bits of feedback/suggestions - if you raise the PR against the `dev` branch we can review properly.

My main feedback is the current behaviour is to try to auto-detect the format and do the right thing. I think the select box you've added should have 'auto-detect' (or similar) as the first and default option.

The other options would then be the same as the HTTP Request node offers:

<img width="463" alt="local_node-red" src="https://user-images.githubusercontent.com/51083/47443072-79582780-d7ab-11e8-8978-f128bc9b344e.png">

**Comment by DurandA** [comment]
I modified the behavior as you suggested. The text is formatted as in `udp in` rather than `http request`:
![mqtt-output](https://user-images.githubusercontent.com/3709037/47654261-b2691100-db8a-11e8-8de0-3d45099ebd0f.png)

Should I insert auto-translated Japanese and Chinese locales or will maintainers do that?

**Comment by knolleary** [comment]
> Should I insert auto-translated Japanese and Chinese locales or will maintainers do that?

No, we have contributors who will take care of that - we don't use any auto-translated messages as they can go badly wrong ;)

**Comment by abrasat** [comment]
Is this feature already implemented in the latest node-red release ?
I need this feature for the MQTT output node, as I need to be able to publish MQTT payloads in binary format.

**Comment by hardillb** [comment]
It was merged to the dev branch as it's a new feature so will be released with the next major release (0.20).

If `msg.payload` is a buffer then it will already be published directly with the current version of the MQTT-out node as described in the the info sidebar
